README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis build
status](https://travis-ci.org/jersme/connectHERE.svg?branch=master)](https://travis-ci.org/jersme/connectHERE)
[![Codecov test
coverage](https://codecov.io/gh/jersme/connectHERE/branch/master/graph/badge.svg)](https://codecov.io/gh/jersme/connectHERE?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/connectHERE)](https://cran.r-project.org/package=connectHERE)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)

# connectHERE

The goal of connectHERE is to connect easier to the HERE REST API.

## Installation

You can install the released version of connectHERE from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("connectHERE")
```

## Code of Conduct

Please note that the ‘connectHERE’ project is released with a
[Contributor Code of Conduct](CODE_OF_CONDUCT.md). By contributing to
this project, you agree to abide by its terms.


NEWS.md
# connectHERE 0.0.0.9001

# connectHERE 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.
* Setup the package structure.


tests/testthat.R
library(testthat)
library(connectHERE)

test_check("connectHERE")


tests/testthat/test.helpers.R
context("API status")

# create test status codes
status_test_200 <- list(status_code = 200)
status_test_401 <- list(status_code = 401)
status_test_402 <- list(status_code = 402)
status_test_300 <- list(status_code = 300)

test_that("api status", {
  expect_true(apiStatusHERE(api_call_object = status_test_200))
  expect_false(apiStatusHERE(api_call_object = status_test_300))
  expect_warning(apiStatusHERE(api_call_object = status_test_300),
                 "API request status: 300 . Please review results, if any, carefully!")
  expect_warning(apiStatusHERE(api_call_object = status_test_401),
                 "Unauthorized: Invalid authentication.")
  expect_false(apiStatusHERE(api_call_object = status_test_401))
  expect_warning(apiStatusHERE(api_call_object = status_test_402),
                 "Forbidden: Incorrect app_code or app_id in the request.")
  expect_false(apiStatusHERE(api_call_object = status_test_402))

})

context("extractDist")

test_route_obj <- list(response.route.summary.distance = 100.25)

test_that("Check inputs", {
  expect_error(extractDist(),
               "No route object specified.")
  expect_error(extractDist(route_object = test_route_obj, unit = "m"),
               "No valid measure for distance specified. Only km or mi is allowed.")
})

test_that("Check outputs", {
  expect_equal(extractDist(route_object = test_route_obj, unit = "km") * .621371,
               extractDist(route_object = test_route_obj, unit = "mi"))
  expect_true(extractDist(route_object = test_route_obj)%%1==0)
  expect_false(extractDist(route_object = test_route_obj, rnd = FALSE)%%1==0)
})

context("createWaypoint")

test_that("Check inputs", {
  expect_error(createWaypoint(lon = 16.37208),
               "Add a lattitude in the lat argument.")
  expect_error(createWaypoint(lat = 48.20849),
               "Add a longitude in the lon argument.")
})

test_that("Check outputs", {
  expect_equal(createWaypoint(48.20849, 16.37208),
               "48.20849,16.37208")
})

context("setURL")

test_that("Check in and output", {
  expect_equal(setURL(type = "routing"),
               "https://route.api.here.com/routing/7.2/calculateroute.json?")
  expect_equal(setURL(type = "geocoding"),
               "https://batch.geocoder.api.here.com/6.2/jobs/")
  expect_error(setURL(type = "xxxx"),
               "Type can only be routing or geocoding.")
  expect_error(setURL(),
               "A valid type must be selected. This can only be routing or geocoding.")
})


tests/testthat/test.setKeys.R
context("setID")

test_that("Check inputs", {
  expect_error(setID(), "Please fill in your HERE REST App ID.")
  expect_warning(setID(id = "xxx"), "Non standard key length.")
  expect_message(setID(id = "xxxxxxxxxxxxxxxxxxxx"), "Key length correct.")
})

test_that("Check outputs", {
  expect_equal(setID(id = "xxxxxxxxxxxxxxxxxxxx"), "xxxxxxxxxxxxxxxxxxxx")
})

context("setCode")

test_that("Check inputs", {
  expect_error(setCode(), "Please fill in your HERE REST App Code.")
  expect_warning(setCode(code = "xxx"), "Non standard key length.")
  expect_message(setCode(code = "xxxxxxxxxxxxxxxxxxxxxx"), "Key length correct.")
})

test_that("Check outputs", {
  expect_equal(setCode(code = "xxxxxxxxxxxxxxxxxxxxxx"), "xxxxxxxxxxxxxxxxxxxxxx")
})


R/setKeys.R
#' Set HERE App ID
#'
#' @param id The project HERE App ID.
#'
#' @return The checked App ID.
#' @export
#'
#' @examples setID(id = "xxxxxxxxxxxxxxxxxxxx")
setID <- function(id) {

  # check id argument
  if (methods::hasArg(id) == FALSE) {
    stop("Please fill in your HERE REST App ID.")
  }

  # check for id length
  if (nchar(id) == 20) {
    message("Key length correct.")
  } else if (nchar(id) != 20) {
    warning("Non standard key length.")
  } else {
    stop("Function error. Please contact package maintainer.")
  }

  # return the keylength
  return(id)
}

#' Set HERE App ID
#'
#' @param code The project HERE App Code.
#'
#' @return The checked App Code.
#' @export
#'
#' @examples setCode(code = "xxxxxxxxxxxxxxxxxxxxxx")
setCode <- function(code) {
  # check id argument
  if (methods::hasArg(code) == FALSE) {
    stop("Please fill in your HERE REST App Code.")
  }

  # check for id length
  if (nchar(code) == 22) {
    message("Key length correct.")
  } else if (nchar(code) != 22) {
    warning("Non standard key length.")
  } else {
    stop("Function error. Please contact package maintainer.")
  }

  # return the keylength
  return(code)
}


R/getRoutes.R
#' Get route information
#'
#' @param url The base url as required to connect to the HERE API. If no value
#' is passed in this parameter the standard parameter from the setURL() is
#' used. See the documentation for the setURL() for more information about
#' the base url.
#' @param app_id App id for the HERE REST API call. See the vignette: Getting
#' Started how to get an App Id.
#' @param app_code App Code for the HERE REST API call. See the vignette: Getting
#' Started how to get an App Code.
#' @param waypoint0 Starting point as a position with in geocoordinates.
#' @param waypoint1 End point as a position with in geocoordinates.
#' @param vehicle Vehicle type for which to calculate the route. Only two
#' options are possible; "truck" or "car". The paramters defaults to "truck".
#'
#' @return A data.frame with all the information about the requested route.
#' @export
#'
#' @examples
#' \dontrun{
#' getRoute(app_id = "xxx", app_code = "xxx", waypoint0 = "52.5,13.45", waypoint1 = "53.5,13.45")
#' }
getRoute <- function(url = setURL(type = "routing"), app_id, app_code, waypoint0, waypoint1, vehicle = "truck") {

  # check for standard arguments
  if (methods::hasArg(url) == FALSE) {
    message("No url in the input. Using the the standard url from setURL().")
  }

  # check for app id and code
  if (methods::hasArg(app_id) == FALSE) {
    stop("No app id in the input.")
  }

  if (methods::hasArg(app_code) == FALSE) {
    stop("No app code in the input.")
  }

  # check for waypoints
  if (methods::hasArg(waypoint0) == FALSE) {
    stop("No waypoint 0 in the input.")
  }

  if (methods::hasArg(waypoint1) == FALSE) {
    stop("No waypoint 1 in the input.")
  }

  # check for vehicle type
  if (vehicle == "truck" | vehicle == "car") {
    message(paste0("Calculation route for: ", vehicle, "."))
  } else {
    stop(paste(vehicle, "is no valid vehicle type selected. Only truck or car is allowed."))
  }

  # call api
  mode_param <- paste0("&mode=fastest;",
                       vehicle,
                       ";traffic:disabled")
  url_call <- paste0(url,
                     "app_id=", app_id,
                     "&app_code=", app_code,
                     "&waypoint0=geo!", waypoint0,
                     "&waypoint1=geo!", waypoint1,
                     mode_param,
                     "&limitedWeight=30.5&height=4.25&shippedHazardousGoods=harmfulToWater")
  tryCatch({
    api_call <- httr::GET(url = url_call)
  })


  # check call status
  call_status <- apiStatusHERE(api_call_object = api_call)

  # get the results out if the API
  results <- as.data.frame(jsonlite::fromJSON(httr::content(api_call, "text"), flatten = TRUE))

  # return the results
  return(results)

}

#' Return delta in distance between truck and car routes
#'
#' @param url The base url as required to connect to the HERE API. If no value
#' is passed in this parameter the standard parameter from the setURL() is
#' used. See the documentation for the setURL() for more information about
#' the base url.
#' @param app_id App id for the HERE REST API call. See the vignette: Getting
#' Started how to get an App Id.
#' @param app_code App Code for the HERE REST API call. See the vignette: Getting
#' Started how to get an App Code.
#' @param waypoint0 Starting point as a position with in geocoordinates.
#' @param waypoint1 End point as a position with in geocoordinates.
#' @param unit The unit of measure of the outputted distance object. Currently
#' only "km" and "mi" are an option.
#' @param rnd Rounds the result. TRUE is the standard setting.
#'
#' @return A list with the truck and car distance
#' @export
#'
#' @examples
#' \dontrun{
#' diffDistTruckCar(
#' app_id = "xxx",
#' app_code = "xxx",
#' waypoint0 = "52.5,13.4",
#' waypoint1 = "54.5,13")
#' }
diffDistTruckCar <- function(url = setURL(type = "routing"), app_id, app_code, waypoint0, waypoint1, unit = "km", rnd = TRUE) {

  # car/truck distance
  tryCatch({
    car_dist <- extractDist(route_object = getRoute(url = url, app_id = app_id, app_code = app_code, waypoint0 = waypoint0, waypoint1 = waypoint1, vehicle = "car"), unit = unit, rnd = rnd)
    truck_dist <- extractDist(route_object = getRoute(url = url, app_id = app_id, app_code = app_code, waypoint0 = waypoint0, waypoint1 = waypoint1, vehicle = "truck"), unit = unit, rnd = rnd)
  })


  # create results list
  results <- list(car = car_dist, truck = truck_dist)

  # return(results)
  return(results)

}

#' Batch input (data.frame) for the diffDistTruckCar() function.
#'
#' @param url The base url as required to connect to the HERE API. If no value
#' is passed in this parameter the standard parameter from the setURL() is
#' used. See the documentation for the setURL() for more information about
#' the base url.
#' @param app_id App id for the HERE REST API call. See the vignette: Getting
#' Started how to get an App Id.
#' @param app_code App Code for the HERE REST API call. See the vignette: Getting
#' Started how to get an App Code.
#' @param df Dataframe where for the distances will be calculated based on the
#' geocoordinates.
#' @param waypoint0 Name of the column for the waypoint 0.
#' @param waypoint1 Name of the column for the waypoint 0.
#'
#' @return The truck and car distances for each each row in the inputted
#' dataframe.
#' @export
#'
#' @examples
#' diffDistTruckCarBatch(
#' df = data.frame(waypoint0 = c("xxx", "yyy"), waypoint1 = c("yyy", "xxx")),
#' app_id = "xxx",
#' app_code = "xxx"
#' )
diffDistTruckCarBatch <- function(url = setURL(type = "routing"), app_id, app_code, df, waypoint0 = "waypoint0", waypoint1 = "waypoint1") {

  # check the inputs
  if (methods::hasArg(df) == FALSE ) {
    stop("Select the data frame containing waypoint 0 and 1")
  }

  if (is.data.frame(df) == FALSE) {
    stop("Dataframe selected is not of type data.frame")
  }

  # create an empty data.frame to add the results later
  distance_all <- data.frame()

  # run the api call over every function
  for (i in seq(1, nrow(df))) {

    # api call

    result <- tryCatch(diffDistTruckCar(app_id = app_id, app_code = app_code, waypoint0 = df[i,waypoint0], waypoint1 = df[i,waypoint1]),
                       warning = function(err_list) {list(car = 99999, truck = 99999)
                       })

    # create the single row data.frame and catch error if the api returns an error code
    distance <- tryCatch(data.frame(car = result$car, truck = result$truck),
                         error = function(err_df){data.frame(car = 99999, truck = 99999)
                      })

    # bind the for loop rows to the data frame
    distance_all <- rbind(distance_all, distance)
  }

  distances <- dplyr::bind_cols(df, distance_all)

  # return  the data frame
  return(distances)
}







R/helpers.R
#' Checks HERE REST API status
#'
#' @param api_call_object The API status object as returned from the httr
#' package.
#'
#' @return The API status in TRUE or FALSE
#' @export
#'
#' @examples
#' \dontrun{
#' apiStatusHERE()
#' }
apiStatusHERE <- function (api_call_object) {

  # check the status
  if (api_call_object$status_code == 200) {
    message("API request status: ok.")
    status <- TRUE
  } else if (api_call_object$status_code == 401) {
    warning("Unauthorized: Invalid authentication.")
    status <- FALSE
  } else if (api_call_object$status_code == 402) {
    warning("Forbidden: Incorrect app_code or app_id in the request.")
    status <- FALSE
  } else {
    warning(paste("API request status:", api_call_object$status_code, ". Please review results, if any, carefully!"))
    status <- FALSE
  }

  # return the status code
  return(status)

}

#' Extracts the distance from a getRoute() result.
#'
#' @param route_object The object as otputted by the getRoute() function.
#' @param unit The unit of measure of the outputted distance object. Currently
#' only "km" and "mi" are an option.
#' @param rnd Rounds the result. TRUE is the standard setting.
#'
#' @return The distance from a getRoute() object in km or mi.
#' @export
#'
#' @examples
#' route_obj <- list(response.route.summary.distance = 100.25)
#' extractDist(route_object = route_obj)
extractDist <- function(route_object, unit = "km", rnd = TRUE) {

  # check for route object
  if (methods::hasArg(route_object) == FALSE) {
    stop("No route object specified.")
  }

  # get the distance from route object.
  dist_km <- route_object$response.route.summary.distance / 1000

  # calculate miles
  if (unit == "mi") {
    dist <- dist_km * .621371
  } else if (unit == "km") {
    dist <- dist_km
  } else {
    stop("No valid measure for distance specified. Only km or mi is allowed.")
  }

  # round
  if (rnd == TRUE) {
    dist <- round(dist)
  }

  # returns the distance value
  return(dist)
}

#' Helper function to create a waypoint from a lattitude and a
#' longitude.
#'
#' @param lat Lattitude coordinates
#' @param lon Longitude coordinates
#'
#' @return A waypoint geocoordinates object as use is the HERE
#' routing API>
#' @export
#'
#' @examples createWaypoint(48.20849, 16.37208)
createWaypoint <- function(lat , lon) {

  # check if lat and lon are present
  if (methods::hasArg(lat) == FALSE) {
    stop("Add a lattitude in the lat argument.")
  }

  if (methods::hasArg(lon) == FALSE) {
    stop("Add a longitude in the lon argument.")
  }

  # create the correct waypoint format
  wp <- paste0(lat, ",", lon)

  # return the waypoint
  return(wp)

}


#' Helper function to set the url in the API calls to the production
#' standards.
#'
#' @param type The type of the API call. Can only be "routing" or "geocoding".
#'
#' @return The standard production url for the selected type.
#' @export
#'
#' @examples
#' setURL(type = "routing")
#' setURL(type = "geocoding")
setURL <- function(type) {

  # check for input
  if (methods::hasArg(type) == FALSE) {
    stop("A valid type must be selected. This can only be routing or geocoding.")
  }

  # routing url
  if (type == "routing") {

    # product standards settings
    route_url <-  "https://route.api.here.com/routing/"
    version <-  "7.2"
    last_part <- "/calculateroute.json?"

    url <- paste0(route_url, version, last_part)

  } else if (type == "geocoding") {

    # product standards settings
    geocode_url <- "https://batch.geocoder.api.here.com/"
    version <-  "6.2"
    last_part <- "/jobs/"

    url <- paste0(geocode_url, version, last_part)

  } else {

    stop("Type can only be routing or geocoding.")

  }

  # return the url
  return(url)

}
